inspect.cpp

```diff
@@ -416,43 +416,91 @@ namespace Sass {
-    string d(ss.str());
-    // store if the value did not equal zero
-    // if after applying precsision, the value gets
-    // truncated to zero, sass emits 0.0 instead of 0
-    bool nonzero = n->value() != 0;
-    size_t decimal = d.find('.');
-    if (decimal != string::npos) {
-      for (size_t i = d.length()-1; d[i] == '0' && i >= decimal; --i) {
-        d.resize(d.length()-1);
+    ss.precision(12);
+    ss << value;
+
+    // check if we got scientific notation in result
+    if (ss.str().find_first_of("e") != string::npos) {
+      ss.clear(); ss.str(string());
+      ss.precision(max(12, precision));
+      ss << fixed << value;
+    }
+
+    const char *it = ss.str().c_str();
+    while (*it && *it != '.' && *it != ',') ++ it;
+    const char *point = it;
+    while (*it && *it == '0') ++ it;
+    bool is_int = (*it == 0);
+    // take a shortcut for integers
+    if (is_int)
+    {
+      ss.clear();
+      ss.str(string());
+      ss.precision(0);
+      ss << fixed << value;
+      res = string(ss.str());
+    }
+    else
+    {
+      const char* end = point; while (*end) ++end;
+      if (end - point - 1 > precision) {
+        ss.clear();
+        ss.str(string());
+        ss.precision(precision);
+        // ss << noshowpoint;
+        ss << fixed << value;
+        res = string(ss.str());
+        size_t pos = res.find_last_not_of("0");
+        bool is_pnt = res[pos] == '.' || res[pos] == ',';
+        res.resize (pos + (is_pnt ? 2 : 1));
+        if (precision == 0) res += ".0";
+      } else {
+        ss.clear();
+        ss.str(string());
+        ss.precision(end - point - 1);
+        ss << fixed << value;
+        res = string(ss.str());
       }
     }
-    if (d[d.length()-1] == '.') d.resize(d.length()-1);
+    // some final cosmetics
+    if (res == "-0.0") res.erase(0, 1);
+    else if (res == "-0") res.erase(0, 1);
+
+    // add unit now
+    res += n->unit();
+
+    // check for a valid unit here
+    // includes result for reporting
     if (n->numerator_units().size() > 1 ||
         n->denominator_units().size() > 0 ||
         (n->numerator_units().size() && n->numerator_units()[0].find_first_of('/') != string::npos) ||
         (n->numerator_units().size() && n->numerator_units()[0].find_first_of('*') != string::npos)
     ) {
-      error(d + n->unit() + " isn't a valid CSS value.", n->pstate());
-    }
-    if (!n->zero() && !in_declaration_list) {
-      if (d.substr(0, 3) == "-0.") d.erase(1, 1);
-      if (d.substr(0, 2) == "0.") d.erase(0, 1);
-    }
-    // remove the leading minus
-    if (d == "-0") d.erase(0, 1);
-    // use fractional output if we had
-    // a value before it got truncated
-    if (d == "0" && nonzero) d = "0.0";
-    // if the precision is 0 sass cast
-    // casts to a float with precision 1
-    if (ctx->precision == 0) d += ".0";
-    // append number and unit
-    append_token(d + n->unit(), n);
+      error(res + " isn't a valid CSS value.", n->pstate());
```

Comment:
Should this be in the output phase? Surely we know if a number in valid in eval?